std/math/cmp/generic.go

```diff
@@ -0,0 +1,91 @@
+// considered to represent integers in range [0, P-1].
+func IsLessUnsigned(api frontend.API, a, b frontend.Variable) frontend.Variable {
+	return isLessRecursive(api, bits.ToBinary(api, a), bits.ToBinary(api, b), false, true)
+}
+
+// IsLessOrEqualUnsigned returns 1 if a <= b, and returns 0 if a > b. The
+// comparison will be unsigned and all field elements will be treated as
+// positive numbers. Therefore, If the underlying field is of prime order P, the
+// elements will be considered to represent integers in range [0, P-1].
+func IsLessOrEqualUnsigned(api frontend.API, a, b frontend.Variable) frontend.Variable {
+	return isLessRecursive(api, bits.ToBinary(api, a), bits.ToBinary(api, b), true, true)
+}
+
+// IsLessBinary compares two binary numbers represented by aBits and bBits. It
+// returns 1 if uint(aBits) < uint(bBits), and returns 0 if uint(aBits) >=
+// uint(bBits). Here, we assume uint(aBits) returns the unsigned integer whose
+// binary representation is aBits.
+func IsLessBinary(api frontend.API, aBits, bBits []frontend.Variable) frontend.Variable {
+	if len(aBits) != len(bBits) {
+		panic("a and b must have the same length")
+	}
+	return isLessRecursive(api, aBits, bBits, false, true)
+}
+
+// IsLessOrEqualBinary compares two binary numbers represented by aBits and
+// bBits. It returns 1 if uint(aBits) <= uint(bBits), and returns 0 if
+// uint(aBits) > uint(bBits). Here, we assume uint(aBits) returns the unsigned
+// integer whose binary representation is aBits.
+func IsLessOrEqualBinary(api frontend.API, aBits, bBits []frontend.Variable) frontend.Variable {
+	if len(aBits) != len(bBits) {
+		panic("a and b must have the same length")
+	}
+	return isLessRecursive(api, aBits, bBits, true, true)
+}
+
+// isLessRecursive compares binary numbers a and b. When useBoundedCmp is false
+// it performs normal bit by bit comparison which defines 2*n multiplication
+// constraints. When useBoundedCmp is true, bit by bit comparison will be used
+// for a few high order bits and the rest of bits will be compared by
+// BoundedComparator. If addition is free, this will define n + 5 constraints
+// when n == api.Compiler().FieldBitLen().
+//
+// acceptEquality determines the return value of the function when a == b.
+func isLessRecursive(api frontend.API,
+	a, b []frontend.Variable,
+	acceptEquality bool, useBoundedCmp bool) frontend.Variable {
+	n := len(a)
+	if n == 0 {
+		if acceptEquality {
+			return 1
+		} else {
+			return 0
+		}
+	}
+	// Interestingly when one of the two numbers is a constant, using bit by bit
+	// comparison will produce 3 fewer constraints than using the BoundedComparator.
+	if useBoundedCmp && n <= api.Compiler().FieldBitLen()-2 &&
+		isNotConstant(api, a[n-1]) && isNotConstant(api, b[n-1]) {
+		diffBound := new(big.Int).Lsh(big.NewInt(1), uint(n))
+		diffBound.Sub(diffBound, big.NewInt(1))
+		comparator := NewBoundedComparator(api, diffBound, false)
```

Comment:
That's a nice feature. However it cannot solve the problem of redundant constraints completely. The parameter could be constrained using custom equations (that could be very complex) or simply it may be a public input. 

Therefore, I think the programmer should always have a way to turn off the binary constraints for parameters. If you think without constraining parameters this function is not safe, we can use the same approach as `bits.ToBinary` and use an enum to turn it off and on.

Reply:
There is `api.MarkBoolean` which records the variable in the table without adding new constraints. So if the user wants to provide bits which they hav itself decomposed somehow, then they have to call `api.MarkBoolean` on the bits.

If possible I would not add options.

Reply:
ah all right, I did not know we had that function.